core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -125,6 +128,21 @@ class PartitionStateMachine(config: KafkaConfig,
     partitionState.filter { case (_, s) => s == state }.keySet.toSet
   }
 
+  private def changeStateTo(partition: TopicPartition, currentState: PartitionState, targetState: PartitionState): Unit = {
+    partitionState.put(partition, targetState)
+    updateControllerMetrics(partition, currentState, targetState)
+  }
+
+  private def updateControllerMetrics(partition: TopicPartition, currentState: PartitionState, targetState: PartitionState) : Unit = {
+    if (!topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic)) {
+      if (currentState != OfflinePartition && targetState == OfflinePartition) {
+        offlinePartitionCount = offlinePartitionCount + 1
+      } else if (currentState == OfflinePartition && targetState != OfflinePartition) {
+        offlinePartitionCount = offlinePartitionCount - 1
```

Comment:
A couple of points on this approach. (1) Currently, if a broker goes offline, the controller calls onReplicasBecomeOffline() which first moves a partition to offline state and then tries to bring the partition to online immediately after. If a leader can be elected, the partition will be transitioned to the online state. With this PR, one potential issue is that there is a very short window in which the offline count could go up and then go down. This can give false alarms in the alerting system. (2) It seems that if a topic is deleted, we don't necessarily go through the partition state transition. We need to update the offline partition count accordingly.

Reply:
(1) You are right that the PartitionStateMachine.offlinePartitionCount can go up and down for some events, e.g. a broker going offline, but the metrics reporter is NOT using this variable directly.
Instead the metrics reporter uses the KafkaController.offlinePartitionCount, which is only updated at the end of processing controller events, so the intermediate value changes won't be exposed to the metrics. 

(2) Currently once a topic is enqueued for deletion in the TopicDeletionManager.topicsToBeDeleted, we don't care about the status of their partitions, and won't count them as offline partitions. I assume we still want to keep the same behavior.
(2.1) With this PR, if partition is online before deletion (offlinePartitionCount counted as 0), then starting the deletion of its topic and maintain the offlinePartitionCount being 0 is the correct behavior.  
(2.2) However if a partition is offline (its offlinePartitionCount counted as 1) when the topic deletion started, then maintaining the value being 1 does not make sense anymore, since we want to stop counting the deleted partition as being offline. This suggests when enqueuing a topic for deletion, we should get the current offline partitions for these topics, and reverse their count in the PartitionStateMachine.offlinePartitionCount variable. 
Does that sound right to you? Thanks!

Reply:
(1) Makes sense. (2) Yes, 2.2 is the case that I was concerned about. Your suggested fix makes sense.